### PR TITLE
sessionctx, types, executor: support encoding and decoding user-defined variables

### LINKED ISSUE #35288: support encoding user vars for session manager
The session manager needs to get all the states for the current session through `show session_states` and restore them on the other session through `set session_states ...`.
This includes user-defined variables.

### PR DESCRIPTION
**Pull request by djshow832**
### What problem does this PR solve?

Issue Number: close #35288

Problem Summary:
This prepares for the session manager: to encode and decode user-defined variables.
It won't affect the user, so it won't be documented.

### What is changed and how it works?

- Marshal and unmarshal Datum, Time, and MyDecimal, which are used in user variables.
- Add `EncodeSessionStates` and `DncodeSessionStates` to `Session`.

### Check List

Tests 

- [x] Unit test

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: sessionctx/variable/session.go [hunk 1]
```diff
@@ -1834,6 +1837,44 @@ func (s *SessionVars) GetTemporaryTable(tblInfo *model.TableInfo) tableutil.Temp
 	return nil
 }
 
+// EncodeSessionStates saves session states into SessionStates.
+func (s *SessionVars) EncodeSessionStates(ctx context.Context, sessionStates *sessionstates.SessionStates) (err error) {
+	// Encode user-defined variables.
+	func() {
+		s.UsersLock.RLock()
```
Comment (bb7133) [inline sessionctx/variable/session.go:1844]: As long as there is no early return for this code block, could you remove the `func()` and `defer`?

**Review by xhebox — APPROVED** [review]
LGTM

**Review by bb7133 — APPROVED** [review]
LGTM

**Pull request merged**